### ISSUE TITLE
Input dtype=float64 causes problems with negloglik

### ISSUE THREAD

**Issue by mckeown12** [tensorflow/probability#342]
The following snippet works as expected if the input dtypes are changed to float32.  It also works if the loss is changed to `'mse'`.

```python
import numpy as np
import tensorflow as tf
from tensorflow.keras.layers import Input
from tensorflow.keras.models import Model
import tensorflow_probability as tfp
tfd = tfp.distributions
print(tf.__version__)
print(tfp.__version__)
#---------------------------------------Mock Data----------
x_data = np.array([[23.0], [24.0], [23.1], [25]], dtype=np.float64)
X={'my_input' :x_data}
y=np.array([[-2.1],[-1.6],[-1.21],[-2.1]], dtype=np.float64)
#---------------------------------------Model definition------
my_input = Input(shape=(1,), dtype='float64', name='my_input')
hidden = tf.keras.layers.Dense(2, activation='relu')(my_input)
p_layer=tfp.layers.DistributionLambda(
      lambda t: tfd.Normal(loc=t[...,0], scale=tf.math.softplus(0.05 * t[...,1])))(hidden)
#--------------------------------------Compile and train------------
negloglik = lambda y, rv_y: -rv_y.log_prob(y)

model=Model(inputs=[my_input], outputs=p_layer)
model.compile(optimizer='adam', loss=negloglik)
model.fit(x=X, y=y, epochs=10, verbose=1, batch_size=4, shuffle=True);
```
As is, we get the following traceback: 

```python traceback
2.0.0-alpha0
0.7.0-dev
Epoch 1/10
---------------------------------------------------------------------------
InvalidArgumentError                      Traceback (most recent call last)
<ipython-input-1-b1db61c7db4d> in <module>
     21 model=Model(inputs=[my_input], outputs=p_layer)
     22 model.compile(optimizer='adam', loss=negloglik)
---> 23 model.fit(x=X, y=y, epochs=10, verbose=1, batch_size=4, shuffle=True);

~/miniconda3/envs/py36/lib/python3.6/site-packages/tensorflow/python/keras/engine/training.py in fit(self, x, y, batch_size, epochs, verbose, callbacks, validation_split, validation_data, shuffle, class_weight, sample_weight, initial_epoch, steps_per_epoch, validation_steps, validation_freq, max_queue_size, workers, use_multiprocessing, **kwargs)
    852           shuffle=shuffle,
    853           initial_epoch=initial_epoch,
--> 854           steps_name='steps_per_epoch')
    855     else:
    856       return training_arrays.fit_loop(

~/miniconda3/envs/py36/lib/python3.6/site-packages/tensorflow/python/keras/engine/training_generator.py in model_iteration(model, data, steps_per_epoch, epochs, verbose, callbacks, validation_data, validation_steps, validation_freq, class_weight, max_queue_size, workers, use_multiprocessing, shuffle, initial_epoch, mode, batch_size, steps_name, **kwargs)
    255 
    256       is_deferred = not model._is_compiled
--> 257       batch_outs = batch_function(*batch_data)
    258       if not isinstance(batch_outs, list):
    259         batch_outs = [batch_outs]

~/miniconda3/envs/py36/lib/python3.6/site-packages/tensorflow/python/keras/engine/training.py in train_on_batch(self, x, y, sample_weight, class_weight, reset_metrics)
   1245           sample_weights=sample_weights,
   1246           reset_metrics=reset_metrics,
-> 1247           output_loss_metrics=self._output_loss_metrics)
   1248     else:
   1249       if not isinstance(K.symbolic_learning_phase(), int):

~/miniconda3/envs/py36/lib/python3.6/site-packages/tensorflow/python/keras/engine/training_eager.py in train_on_batch(model, inputs, targets, sample_weights, reset_metrics, output_loss_metrics)
    293           sample_weights=sample_weights,
    294           training=True,
--> 295           output_loss_metrics=output_loss_metrics))
    296   if not isinstance(outs, list):
    297     outs = [outs]

~/miniconda3/envs/py36/lib/python3.6/site-packages/tensorflow/python/keras/engine/training_eager.py in _process_single_batch(model, inputs, targets, output_loss_metrics, sample_weights, training)
    233               output_loss_metrics=output_loss_metrics,
    234               sample_weights=sample_weights,
--> 235               training=training))
    236       if total_loss is None:
    237         raise ValueError('The model cannot be run '

~/miniconda3/envs/py36/lib/python3.6/site-packages/tensorflow/python/keras/engine/training_eager.py in _model_loss(model, inputs, targets, output_loss_metrics, sample_weights, training)
    150           current_loss_reduction = loss_fn.reduction
    151           loss_fn.reduction = losses_utils.ReductionV2.NONE
--> 152           weighted_losses = loss_fn(targets[i], outs[i], sample_weight=weights)
    153           loss_fn.reduction = current_loss_reduction
    154 

~/miniconda3/envs/py36/lib/python3.6/site-packages/tensorflow/python/keras/losses.py in __call__(self, y_true, y_pred, sample_weight)
     94     with ops.name_scope(scope_name, format(self.__class__.__name__),
     95                         (y_pred, y_true, sample_weight)):
---> 96       losses = self.call(y_true, y_pred)
     97       return losses_utils.compute_weighted_loss(
     98           losses, sample_weight, reduction=self.reduction)

~/miniconda3/envs/py36/lib/python3.6/site-packages/tensorflow/python/keras/losses.py in call(self, y_true, y_pred)
    156       Loss values per sample.
    157     """
--> 158     return self.fn(y_true, y_pred, **self._fn_kwargs)
    159 
    160   def get_config(self):

<ipython-input-1-b1db61c7db4d> in <lambda>(y, rv_y)
     17       lambda t: tfd.Normal(loc=t[...,0], scale=tf.math.softplus(0.05 * t[...,1])))(hidden)
     18 #--------------------------------------Compile and train------------
---> 19 negloglik = lambda y, rv_y: -rv_y.log_prob(y)
     20 
     21 model=Model(inputs=[my_input], outputs=p_layer)

~/miniconda3/envs/py36/lib/python3.6/site-packages/tensorflow_probability/python/distributions/distribution.py in log_prob(self, value, name)
    834         values of type `self.dtype`.
    835     """
--> 836     return self._call_log_prob(value, name)
    837 
    838   def _call_prob(self, value, name, **kwargs):

~/miniconda3/envs/py36/lib/python3.6/site-packages/tensorflow_probability/python/distributions/distribution.py in _call_log_prob(self, value, name, **kwargs)
    817           value, name="value", preferred_dtype=self.dtype)
    818       if hasattr(self, "_log_prob"):
--> 819         return self._log_prob(value, **kwargs)
    820       if hasattr(self, "_prob"):
    821         return tf.math.log(self._prob(value, **kwargs))

~/miniconda3/envs/py36/lib/python3.6/site-packages/tensorflow_probability/python/distributions/normal.py in _log_prob(self, x)
    187 
    188   def _log_prob(self, x):
--> 189     return self._log_unnormalized_prob(x) - self._log_normalization()
    190 
    191   def _log_cdf(self, x):

~/miniconda3/envs/py36/lib/python3.6/site-packages/tensorflow_probability/python/distributions/normal.py in _log_unnormalized_prob(self, x)
    202 
    203   def _log_unnormalized_prob(self, x):
--> 204     return -0.5 * tf.square(self._z(x))
    205 
    206   def _log_normalization(self):

~/miniconda3/envs/py36/lib/python3.6/site-packages/tensorflow_probability/python/distributions/normal.py in _z(self, x)
    227     """Standardize input `x` to a unit normal."""
    228     with tf.compat.v1.name_scope("standardize", values=[x]):
--> 229       return (x - self.loc) / self.scale
    230 
    231   def _inv_z(self, z):

~/miniconda3/envs/py36/lib/python3.6/site-packages/tensorflow/python/ops/math_ops.py in binary_op_wrapper(x, y)
    861     with ops.name_scope(None, op_name, [x, y]) as name:
    862       if isinstance(x, ops.Tensor) and isinstance(y, ops.Tensor):
--> 863         return func(x, y, name=name)
    864       elif not isinstance(y, sparse_tensor.SparseTensor):
    865         try:

~/miniconda3/envs/py36/lib/python3.6/site-packages/tensorflow/python/ops/gen_math_ops.py in sub(x, y, name)
  10620       else:
  10621         message = e.message
> 10622       _six.raise_from(_core._status_to_exception(e.code, message), None)
  10623   # Add nodes to the TensorFlow graph.
  10624   _, _, _op = _op_def_lib._apply_op_helper(

~/miniconda3/envs/py36/lib/python3.6/site-packages/six.py in raise_from(value, from_value)

InvalidArgumentError: cannot compute Sub as input #1(zero-based) was expected to be a float tensor but is a double tensor [Op:Sub] name: model/distribution_lambda/Normal/log_prob/standardize/sub/
```

I have encountered similar issues when using the SinhArcsinh distribution.  I get errors like `Inconsistent values for attr 'T' DT_FLOAT vs. DT_DOUBLE` and also  `TypeError: Input had dtype <dtype: 'float32'> but expected <dtype: 'float64'>` when passing `np.float64`s to the Input.

**Comment by csuter** [comment]
I get shape errors when I try running this as written -- did you see those previously and fix somehow, but maybe this snippet doesn't reflect the fix?

**Comment by csuter** [comment]
Meanwhile, I don't expect it to but does setting the keras default floatx type help? (c.f., https://stackoverflow.com/questions/50579167/change-keras-floatx-default-float-type-in-python)

**Comment by jburnim** [comment]
It looks like you are hitting a bug in tf-2.0.0-alpha0, which has been fixed in tf-nightly and tf-nightly-2.0-preview.  I created a small Colab at https://colab.research.google.com/drive/1SJISh2_Qz31lguu4XgC4jMrlVRAkHgOg to reproduce this error in tf-2.0.0-alpha0 without any use of TFP:
```
x_data = np.array([[23.0], [24.0], [23.1], [25]], dtype=np.float64)
X={'my_input' :x_data}
y=np.array([[-2.1],[-1.6],[-1.21],[-2.1]], dtype=np.float64)

my_input = Input(shape=(1,), dtype=tf.float64, name='my_input')
hidden = tf.keras.layers.Dense(2, activation='relu')(my_input)
output = tf.keras.layers.Dense(1, activation='relu')(hidden)

loss = lambda y_true, y_pred: tf.square(y_true - y_pred)

model = Model(inputs=my_input, outputs=output)
model.compile(optimizer='adam', loss=loss)
model.fit(x=X, y=y, epochs=10, verbose=1, batch_size=4, steps_per_epoch=1, shuffle=True)
```
Can you please let us know if the error goes away if you either?:

 (1) Switch to using tf-nightly and tf-nightly-2.0-preview.

 (2) Add an explicit `-rv_y.log_prob(tf.cast(y, tf.float64))` to work around the bug in tf-2.0.0-alpha0.

**Comment by axch** [comment]
Based on the above discussion, I am going to close this issue as resolved.  @mckeown12, please reopen if you are still experiencing difficulties.
